### PR TITLE
Week10

### PR DESCRIPTION
**Pull request by oishikg**

### CODE REVIEW

File: week10/hw10/AParser.hs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+------------------------------------------------------------
+
+{------------------------- QN 1 -------------------------}
+
+-- First, you’ll need to implement a Functor instance for Parser.
+-- Hint: You may find it useful to implement a function
+-- first :: (a -> b) -> (a,c) -> (b,c)
+
+first :: (a -> b) -> (a,c) -> (b,c)
+first f (a , c) = (f a , c)
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: It's nice to implement such function manually. But in practice we usually use `first` from the `Data.Bifunctor` module.

* https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bifunctor.html#v:first

File: week10/hw10/AParser.hs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+first :: (a -> b) -> (a,c) -> (b,c)
+first f (a , c) = (f a , c) 
+
+instance Functor Parser where
+  fmap f p = Parser g
+    where
+      g = \str ->
+            case (runParser p) str of
+              Nothing -> Nothing
+              Just p -> Just (first f p)
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: I think that this implementation of `fmap` is a little bit verbose. Would be nice if you can implement it simple in point-free style, using the fact that `Maybe` also has a `Functor` instance.

File: week10/hw10/AParser.hs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+-- *AParser> runParser abParser "abcdef"
+-- Just ((’a’,’b’),"cdef")
+-- *AParser> runParser abParser "aebcdf"
+-- Nothing
+
+
+
+abParser :: Parser (Char, Char)
+abParser =  ( , ) <$> (char 'a') <*> (char 'b')
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: Tuple constructor as an operator is usually used without spaces
```suggestion
abParser =  (,) <$> (char 'a') <*> (char 'b')
```

File: week10/hw10/AParser.hs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+-- which acts in the same way as abParser but returns () instead of
+-- the characters ’a’ and ’b’.
+
+-- *AParser> runParser abParser_ "abcdef"
+-- Just ((),"cdef")
+-- *AParser> runParser abParser_ "aebcdf"
+-- Nothing
+
+abParser_ =  (\_ _ -> ()) <$> (char 'a') <*> (char 'b')
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: It would be nice if you could reuse your `abParser` here to not duplicate code

File: week10/hw10/AParser.hs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+-- • empty represents the parser which always fails.
+-- • p1 <|> p2 represents the parser which first tries running p1. If
+-- p1 succeeds then p2 is ignored and the result of p1 is returned.
+-- Otherwise, if p1 fails, then p2 is tried instead.
+
+-- Hint: there is already an Alternative instance for Maybe which you
+-- may find useful
+
+instance Alternative Parser where
+  empty = Parser {runParser = \str -> Nothing}
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: You can use `const` function here

File: week10/hw10/AParser.hs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  
+  p1 <|> p2 = Parser f
+    where
+      f = \str ->
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: This function can be written simple if you use `Alternative` instance for `Maybe`

File: week10/hw10/AParser.hs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+-- *Parser> runParser intOrUppercase "foo"
+-- Nothing
+
+-- Next week, we will use your parsing framework to build a more
+-- sophisticated parser for a small programming language!
+
+intOrUppercase :: Parser ()
+intOrUppercase = (toEmpty <$> posInt) <|> (toEmpty <$> satisfy isUpper)
+  where
+    toEmpty _ = ()
```
Comment (chshersh) [inline week10/hw10/AParser.hs]: Have a look at the `void` function:

https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Monad.html#v:void

**Review by chshersh — COMMENTED** [review]
Looks nice, but many functions can be improved using common idioms and standard functions from `base`

File: week10/hw10/AParser.hs (hunk 1 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Noted! I've removed the definition and imported the module.

File: week10/hw10/AParser.hs (hunk 2 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Done!

File: week10/hw10/AParser.hs (hunk 3 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Made this change

File: week10/hw10/AParser.hs (hunk 4 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Used the `const` function to do so

File: week10/hw10/AParser.hs (hunk 5 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Implemented!

File: week10/hw10/AParser.hs (hunk 6 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Yes, it is much more concise. Thanks for pointing this out

File: week10/hw10/AParser.hs (hunk 7 above)
Comment (oishikg) [reply week10/hw10/AParser.hs]: Rewrote the function using void!

File: week10/hw10/AParser.hs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+-- which acts in the same way as abParser but returns () instead of
+-- the characters ’a’ and ’b’.
+
+-- *AParser> runParser abParser_ "abcdef"
+-- Just ((),"cdef")
+-- *AParser> runParser abParser_ "aebcdf"
+-- Nothing
+
+abParser_ = const () <$> abParser
```
Comment (chshersh) [inline week10/hw10/AParser.hs:146]: `const () <$> x` is just `void x`

**Review by chshersh — APPROVED** [review]
Looks much cleaner now!

**Pull request merged**